inst/tutorials/practical0-1/practical0-1.Rmd

```diff
@@ -0,0 +1,444 @@
+`addtwonumberstogether`, whereas `add_two_numbers_together` is easier.
+However, there are boring reasons to do with how R works why it's a bad idea to
+write R functions with dots in their names (even though many functions in R
+already have dots). As a result, in these practicals, I insist that you don't
+uses dots in function names when you write them.
+
+However, dots are generally okay in variables names, so to make it easy to
+distinguish normal variables from functions, I choose to put dots in variable
+names (as `one.more`). You can do whatever you like.
+
+Here's a function that takes two arguments and adds them together (yes, it
+is just reinventing `+`!):
+
+```{r rename, exercise.cap = "Exercise 2", exercise=TRUE}
+add.up <- function(first, second) {
+  added_together <- first + second
+  added_together
+}
+
+add.up(10, 30)
+```
+
+```{r rename-solution}
+add_up <- function(first, second) {
+  added.together <- first + second
+  added.together
+}
+
+add_up(10, 30)
+```
+
+```{r rename-check}
+grade_this_code()
+```
+
+Rename the function and variables in *Exercise 2* so that it satisfies my
+naming rules.
+
+### Not reusing names
+
+In *Exercise 1*, we showed that variables used in functions don't affect what
+happens outside the function. Unfortunately, the reverse isn't true. Look at
+this function:
+
+```{r global, exercise.cap = "Exercise 3", exercise=TRUE}
+missing_add <- function(first) {
+  first + second
+}
+
+missing_add(10)
+```
+
+This function was intended to be like `add_up()` in *Exercise 2*, but I
+accidentally forgot to add the second argument. Fortunately if you run it, it
+will give an error (try it). However, what happens if you have defined
+a variable called `second` elsewhere in your script? Try adding the following
+at line 4:
+
+```{r, eval=FALSE}
+second <- 1
+```
+
+`missing_add(10)` should no longer give an error, and return `11`. Try changing
+`second` to 2 - `missing_add(10)` should now return `12`. That's not good news -
+what the function does depends on what else you may have done in your script.
+When you call a function it should always do the same if you give it the same
+arguments.
+
+This gets even worse when you account for our ability to misspell things:
+
+```{r misspelling, exercise.cap = "Exercise 3", exercise=TRUE}
+misspelled_add <- function(first, second) {
+  resuIt <- first + second
+  result
+}
+
+result <- -5
+misspelled_add(10, 20)
+misspelled_add(1, 2)
+```
+
+<div id="misspelling-hint">
+**Hint:** If you can't see the problem see how the word `result` is spelled in
+different places.
+</div>
+
+You should find that `misspelled_add()` gives the wrong answer even though the
+code looks right (if you don't look too closely). This problem can (and does!)
+cause hours of agony to some students every year on this course. As a result,
+we insist on two things in your submitted practicals:
+
+1. You use variable names inside functions that you don't use outside in your
```

Comment:
"You use different variable names in your functions than in your scripts, wherever possible"

Also, can we say "variable names" instead of "object names"? I changed this a couple of times in some of the exercises when objects seemed the focus, but haven't done a full sweep.

Reply:
I'm not sure quite what you mean about the objects, maybe we need to talk about that later. Anyway, I've accepted your edit.